dali/test/dali_test_checkpointing.h

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_TEST_DALI_TEST_CHECKPOINTING_H_
+#define DALI_TEST_DALI_TEST_CHECKPOINTING_H_
+
+#include <gtest/gtest.h>
+
+#include <string>
+#include <utility>
+#include <vector>
+#include <memory>
+
+#include "dali/test/dali_test.h"
+#include "dali/pipeline/operator/checkpointing/checkpoint.h"
+
+namespace dali {
+
+/**
+ * @brief Pipeline wrapper that allows deep copying in checkpointing tests.
+*/
+class PipelineWrapper {
+ public:
+  using OutputsType = std::vector<std::pair<std::string, std::string>>;
+
+  inline PipelineWrapper(int batch_size, OutputsType outputs)
+    : batch_size_(batch_size),
+      pipe_(new Pipeline(batch_size, 1, 0)),
+      outputs_(outputs) {}
+
+  inline void AddOperator(OpSpec spec) {
+    std::string identifier = std::to_string(next_identifier_++);
+    AddOperator(spec, identifier);
+  }
+
+  inline void AddOperator(OpSpec spec, std::string identifier) {
+    pipe_->AddOperator(spec, identifier);
+    ops_.emplace_back(spec, identifier);
+  }
+
+  inline void Build() {
+    pipe_->Build(outputs_);
+  }
+
+  /**
+   * @brief Make a deep copy of a pipeline, transferring state by checkpointing.
+   */
+  inline PipelineWrapper CopyByCheckpointing() {
+    PipelineWrapper clone(batch_size_, outputs_);
+    for (const auto &[spec, id] : ops_)
+      clone.AddOperator(spec, id);
+    clone.Build();
+
+    for (const auto &[spec, id] : ops_) {
+      // Only save/restore state for operators with checkpointing enabled.
+      bool checkpointing;
+      if (!spec.TryGetArgument(checkpointing, "checkpointing") || !checkpointing)
+        continue;
```

Comment:
Is this needed?

Reply:
Then it removes the possibility to get input from an stateless external source in the tested pipeline.

I don't use it in tests, and it is easy to mess up testing this way, so I can remove it.